modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/DynamicOptionsParameterDeclarationEnricher.java

```diff
@@ -0,0 +1,254 @@
+import org.mule.runtime.module.extension.internal.loader.java.property.OptionsResolverFactoryModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.OptionsResolverFactoryModelProperty.OptionsResolverFactoryModelPropertyBuilder;
+import org.mule.runtime.module.extension.internal.loader.java.property.ParameterGroupModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.type.ExtensionParameter;
+import org.mule.runtime.module.extension.internal.loader.java.type.FieldElement;
+import org.mule.runtime.module.extension.internal.loader.java.type.runtime.ParameterizableTypeWrapper;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * {@link DeclarationEnricher} implementatio
+ *
+ * @since 4.0
+ */
+public class DynamicOptionsParameterDeclarationEnricher extends AbstractAnnotatedDeclarationEnricher {
+
+  private ClassTypeLoader typeLoader;
+
+  @Override
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    Optional<ImplementingTypeModelProperty> implementingType =
+        extractImplementingTypeProperty(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+
+    typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader(currentThread().getContextClassLoader());
+
+    if (implementingType.isPresent()) {
+      new IdempotentDeclarationWalker() {
+
+        @Override
+        public void onSource(SourceDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+
+        @Override
+        public void onOperation(OperationDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+
+        @Override
+        protected void onConfiguration(ConfigurationDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+
+        @Override
+        protected void onConnectionProvider(ConnectionProviderDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+      }.walk(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+    }
+  }
+
+  /**
+   *
+   * @param declaration
+   * @param enrichableModel
+   */
+  private void enrichContainerModel(ParameterizedDeclaration<?> declaration, Consumer<ModelProperty> enrichableModel) {
+    List<ParameterDeclaration> allParameters = declaration.getAllParameters();
+
+    Map<ParameterGroupDeclaration, Class<? extends OptionsResolver>> dynamicGroupOptions =
+        introspectParameterGroups(declaration.getParameterGroups());
+    Map<ParameterDeclaration, Class<? extends OptionsResolver>> dynamicOptions = introspectParameters(allParameters);
```

Comment:
`allParameters` include the parameters that where already processed in the previous round of `parameterGroups`. Iterate once over all the groups and process each of them accordingly. Just consider duplication between groups for sources callbacks